### DISCUSSION TITLE
When I am updating a signal in `createEffect` I want to execute a part of component which is inside a component body

### DISCUSSION THREAD

**Discussion by wajeshubham** [solidjs/solid#1100] [Q&A] [answered]
### Describe the bug

I am creating a dynamic component to secure my routes but when I am updating the local signal `setIsValid` in `createEffect` the changes are not causing the rerun of changed signal.

```jsx
import { Navigate } from "solid-app-router"
import { Component, createEffect, createSignal, JSX } from "solid-js"
import { auth } from "../app/authStore"

const PrivateRoute: Component<{ children: JSX.Element }> = (props) => {
  const [isValid, setIsValid] = createSignal<boolean | undefined>(undefined)
  createEffect(() => {
    if (auth.isAuthenticated) {
      setIsValid(true)
      return
    }
    setIsValid(false)
  })

  if (isValid() === undefined) {
    return <>Loading...</>
  }

  // I want to re execute the following logic once isValid() is changed
  // I assume because solid js does not re render the whole component this might not be the right way to implement this logic
  // It would be a great help if you give some input on how to implement the same

  if (isValid()) {
    return props.children
  }

  return <Navigate href="/login" />
}

export default PrivateRoute

```

**SOME CONTEXT BEFOR YOU SEE THE ISSUE:** As per my knowledge the solidJS works on the subscription principle and it doesn't re-render the whole component like ReactJs does. So because the `PrivateRoute` component is not getting re-rendered, the `if` statements are not getting executed multiple times hence causing it to not return the component when I am changing the `isValid` signal 

It would be a great help if you give us some pointers on how to implement the same.

### Your Example Website or App

https://stackblitz.com/edit/vitejs-vite-8qywzr

### Steps to Reproduce the Bug or Issue

1. Go to `src/components/PrivateRoute.tsx` and see the comments I have added in `createEffect`.
2. signal `isValid` is staying `undefined` even after setting it to true

### Expected behavior

It should rerun the redirection logic. That is when user logged out and I am in `PrivateRoute` page user should get redirected to login page

### Platform

- MacOS
- Chrome

**Comment by edemaine** [comment, marked as answer]
You can use [the `<Show>` component](https://www.solidjs.com/docs/latest/api#%3Cshow%3E) and the following pattern to get reactive toggling:

```jsx
  return (
    <Show when={isValid()} fallback={<Navigate href="/login" />}>
      {props.children}
    </Show>
  );
```

In the future, you might want to ask this sort of question the Discord server, in the #help channel.

**Comment by wajeshubham** [comment]
```jsx
import { Navigate, Outlet } from "solid-app-router"
import { Component, Show } from "solid-js"
import { auth } from "../app/authStore"

const PrivateRoute: Component = () => {
  return (
    <Show when={auth.isAuthenticated} fallback={<Navigate href="/login" />}>
      <Outlet />
    </Show>
  )
}

export default PrivateRoute
```
Is this a right approach too? `auth.isAuthenticated` is my store created using `createStore`
